src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -74,13 +106,242 @@ public class ConsensusManager
                 utxoHash = await this.consensusRules.RewindAsync().ConfigureAwait(false);
             }
 
-            this.concurrentChain.SetTip(this.Tip);
+            this.chainedHeaderTree.Initialize(this.Tip, !blockStoreDisabled);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A list of headers are presented from a given peer,
+        /// we'll attempt to connect the headers to the tree and if new headers are found they will be queued for download.
+        /// </summary>
+        /// <param name="peerId">The peer that is providing the headers.</param>
+        /// <param name="headers">The list of new headers.</param>
+        /// <returns>The last chained header that is connected to the tree.</returns>
+        public ChainedHeader HeadersPresentedAsync(int peerId, List<BlockHeader> headers)
+        {
+            this.logger.LogTrace("({0}:{1},{2}.{3}:{4})", nameof(peerId), peerId, nameof(headers), nameof(headers.Count), headers.Count);
 
-            this.lookaheadBlockPuller.SetLocation(this.Tip);
+            ConnectNewHeadersResult newHeaders = null;
 
-            this.chainedHeaderTree.Initialize(this.Tip, !blockStoreDisabled);
+            lock (this.treeLock)
+            {
+                newHeaders = this.chainedHeaderTree.ConnectNewHeaders(peerId, headers);
+                this.blockPuller.NewTipClaimed();
+            }
+
+            if (newHeaders?.DownloadTo != null)
+            {
+                this.DownloadBlocks(newHeaders.ToHashList(), this.ProcessDownloadedBlock);
+            }
+
+            this.logger.LogTrace("(-):'{0}'", newHeaders?.Consumed);
+            return newHeaders?.Consumed;
+        }
+
+        /// <summary>
+        /// A peer was disconnected, clean all its claimed headers that are above our consensus tip.
+        /// Processes any remaining blocks to download.
+        /// </summary>
+        /// <param name="peerId">The peer that is providing the headers.</param>
+        public void PeerDisconnected(int peerId)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(peerId), peerId);
+
+            lock (this.treeLock)
+            {
+                this.chainedHeaderTree.PeerDisconnected(peerId);
+                this.blockPuller.PeerDisconnected(peerId);
+                this.ProcessDownloadQueueLocked();
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A callback that is triggered when a block was downloaded.
```

Comment:
This callback is called only when CM itself requested this block. It's not called when a block requested by other component is downloaed